crates/hir-ty/src/infer.rs

```diff
@@ -905,8 +873,68 @@ impl<'a> InferenceContext<'a> {
             TypeNs::SelfType(impl_id) => {
                 let generics = crate::utils::generics(self.db.upcast(), impl_id.into());
                 let substs = generics.placeholder_subst(self.db);
-                let ty = self.db.impl_self_ty(impl_id).substitute(Interner, &substs);
-                self.resolve_variant_on_alias(ty, unresolved, mod_path)
+                let mut ty = self.db.impl_self_ty(impl_id).substitute(Interner, &substs);
+
+                let Some(mut remaining_idx) = unresolved else {
+                    return self.resolve_variant_on_alias(ty, None, mod_path);
+                };
+
+                let mut remaining_segments = path.segments().skip(remaining_idx);
+
+                // We need to try resolving unresolved segments one by one because each may resolve
+                // to a projection, which `TyLoweringContext` cannot handle on its own.
+                while !remaining_segments.is_empty() {
+                    let resolved_segment = path.segments().get(remaining_idx - 1).unwrap();
+                    let current_segment = remaining_segments.take(1);
```

Comment:
There are a bunch of `unwrap` in this function, which I think it is possible to remove by defining some windows like thing for `PathSegments` (or by using `Itertools::tuple_windows`) and a function which goes from a `&PathSegment` to a `&PathSegments` (similar to `slice::from_ref`).

Reply:
we've got these path segment unwraps in more places (our API for paths isn't really that great currently)

Reply:
Okay, so we can do it in a follow up.